File: 53-maximum-subarray/53-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
  let l = 0;
  let r = 0;
  const arrLen = nums.length;
  let sum;
  let maxSum = -Infinity;
  
  while (r < arrLen) {
    if (l === r) {
      sum = nums[l];
    } else {
      sum = sum + nums[r];
    }
    
    r ++;
      
    if (sum < 0) l = r;
    maxSum = Math.max(maxSum, sum);
    
  }
  
    
  return maxSum;
};

File: 53-maximum-subarray/53-maximum-subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
    sum_at_left = 0
    max_sum = nums[0]
    
    nums.each do |num|
        sum_at_left = 0 if sum_at_left < 0
        sum_at_left += num
        max_sum = [sum_at_left, max_sum].max
    end
    
    max_sum
end